### PR TITLE
Add hook to update io.input.stream to sys.stdin if it is None.

### PR DESCRIPTION
**Pull request by chadac**
# Pull Request Check List

Resolves: #5633 

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

### CODE REVIEW

File: src/poetry/console/commands/command.py [hunk 1]
```diff
@@ -17,6 +21,17 @@ class Command(BaseCommand):  # type: ignore[misc]
...
 
+    def run(self, io: IO) -> int:
+        """
+        Temporarily fix for issue where io.input.stream is unset by cleo.
+        """
+        # FIXME: Remove method after upstream merge for cleo
+        # https://github.com/sdispater/cleo/pull/135
+        if io.input.stream is None:
+            io.input.set_stream(sys.stdin)
+        status_code: int = super().run(io)
```
Comment (Secrus) [inline src/poetry/console/commands/command.py:32]: Is that extra variable needed? can't we just `return super().run(io)`?

Comment (chadac) [reply src/poetry/console/commands/command.py:32]: mypy complains about the type for `super.run()` being `Any` if I return it directly, so I wrapped it into an explicit cast here. I could swap to a mypy flag if that works better, though.

Comment (Secrus) [reply src/poetry/console/commands/command.py:32]: No, lets leave like this.

**Review by Secrus — APPROVED** [review]

**Review by Secrus — APPROVED** [review]

**Comment by abn** [comment]
Is this something we can add a meaningful test for?

**Comment by chadac** [comment]
@abn 

I could create a new `test_command.py` with these tests or add them to an existing file:

```
def test_command_with_no_io_stream(mocker: MockerFixture):
    command = Command()
    mocker.patch.object(
        BaseCommand,
        "run",
        return_value=[1]
    )
    io = IO(input=Input(), output=None, error_output=None)
    io.input.set_stream(None)
    command.run(io)
    assert io.input.stream == sys.stdin


def test_command_with_existing_io_stream(mocker: MockerFixture):
    command = Command()
    mocker.patch.object(
        BaseCommand,
        "run",
        return_value=[1]
    )
    io = IO(input=Input(), output=None, error_output=None)
    io.input.set_stream(sys.stdout)
    command.run(io)
    assert io.input.stream == sys.stdout
```

What would y'all recommend?

**Comment by abn** [comment]
I was wondering if we could create a more functional test. Similar to one of the other interactive command test cases.

**Comment by chadac** [comment]
Sure, lemme take a crack at that. There currently aren't any tests for `cache clear` at the moment, I can see if I can get it fully covered in the integration tests.

**Comment by chadac** [comment]
I've gone ahead and added some tests for the `cache clear` command, but incidentally it actually doesn't capture the issue this PR fixes. The main issue is that the `CommandTester` in cleo don't actually trigger the problematic code in `cleo.application.Application:L425`. It's possible that this could be caught by migrating to using `ApplicationTester` instead, although I don't know what y'all think about that.

**Comment by neersighted** [comment]
> I've gone ahead and added some tests for the `cache clear` command, but incidentally it actually doesn't capture the issue this PR fixes. The main issue is that the `CommandTester` in cleo don't actually trigger the problematic code in `cleo.application.Application:L425`. It's possible that this could be caught by migrating to using `ApplicationTester` instead, although I don't know what y'all think about that. It would be a divergence from what most of the other commands do.

I would rather have a good regression test for this, even if it's a bit divergent from the other tests.

**Comment by chadac** [comment]
Sounds good, I've gone ahead and swapped it over. I I did have to add one attribute to the `Application` class in order to ensure that the "original" input object is sufficiently preserved, so that the command code can property specify either sys.stdin or the proper test StringIO stream. I'm open to other ideas, though. It's technically less brittle now as the hardcoded sys.stdin is gone, but technically it should be unneeded when cleo merges in this PR and has the same tests.

**Comment by branchvincent** [comment]
@chadac we plan to have a new `cleo` release soon with your merged fix. Could we update this pr to temporarily depend on cleo's `master` branch to verify your tests also pass here? Then we can update again once it's released

**Comment by neersighted** [comment]
The Cleo release is now in master -- this can just be converted to a regression test only.

**Review by neersighted — APPROVED** [review]

**Pull request merged**